torch/_inductor/fx_passes/decompose_mm.py

```diff
@@ -0,0 +1,569 @@
+    # out = torch.empty_strided(
+    #     size=orig_out.shape,
+    #     stride=orig_out.stride(),
+    #     dtype=orig_out.dtype,
+    #     device=orig_out.device,
+    # )
+
+    rs_outs = []
+    for a_i, b_i in zip(mm_i_a_args, mm_i_b_args):
+        mm_i = torch.ops.aten.mm(a_i, b_i)
+        for op in post_mm_ops:
+            mm_i = op.apply(mm_i, num_chunks)
+        if scatter_dim != 0:
+            mm_i_chunks = mm_i.chunk(orig_split_num_chunks, dim=scatter_dim)
+            cat = torch.cat(mm_i_chunks)
+            mm_i = cat
+        w = torch.ops._c10d_functional.reduce_scatter_tensor(
+            mm_i, reduce_op, group_size, group_name
+        )
+        rs_out_i = torch.ops._c10d_functional.wait_tensor(w)
+        rs_outs.append(rs_out_i)
+
+    if scatter_dim == 0:
+        # B column wise split
+        rs_out = torch.cat(rs_outs, dim=-1)
+    else:
+        # A row wise split
+        rs_out = torch.cat(rs_outs)
+    return (rs_out,)
+
+
+def _mm_split_cat_rs_fn_view3d(
+    a: torch.Tensor,
+    b: torch.Tensor,
+    num_chunks: int,
+    orig_out: torch.Tensor,
+    reduce_op: str,
+    group_size: int,
+    group_name: str,
+    scatter_dim: int,
+    orig_split_num_chunks: int,
+    post_mm_ops,
+):
+    assert isinstance(num_chunks, int)
+    mm_i_a_args = [a] * num_chunks
+    mm_i_b_args = [b] * num_chunks
+    a_flat = a.flatten(0, -2)
+    a_flat_chunks = a_flat.chunk(num_chunks)
+    mm_i_a_args = a_flat_chunks
+
+    rs_outs = []
+    for a_i, b_i in zip(mm_i_a_args, mm_i_b_args):
+        mm_i = torch.ops.aten.mm(a_i, b_i)
+        for op in post_mm_ops:
+            mm_i = op.apply(mm_i, num_chunks)
+
+        mm_i_chunks = mm_i.chunk(orig_split_num_chunks, dim=scatter_dim)
+        cat = torch.cat(mm_i_chunks)
+        mm_i = cat
+
+        w = torch.ops._c10d_functional.reduce_scatter_tensor(
+            mm_i, reduce_op, group_size, group_name
+        )
+        rs_out_i = torch.ops._c10d_functional.wait_tensor(w)
+        rs_outs.append(rs_out_i)
+
+    rs_out = torch.cat(rs_outs)
+    return (rs_out,)
+
+
+def _size_hint(s: sympy.Expr) -> int:
+    hint = size_hint(s)
+    if hint is not None:
+        return hint
+    return 0
+
+
+def _is_contiguous(t) -> bool:
+    return t.is_contiguous(memory_format=torch.contiguous_format)
+
+
+def _select_optimal_chunk_count(
+    size_before_split: int,
+    num_chunks: list[int],
+    min_size_after_split: int,
+    additional_check=None,
+):
+    """Pick the largest chunk count that satisfies minimum size constraint.
+
+    Returns the greatest number from num_chunks such that
+    size_before_split // chunk_count >= min_size_after_split.
+
+    Args:
+        size_before_split: Total size to be split
+        num_chunks: List of candidate chunk counts
+        min_size_after_split: Minimum size each chunk must have
+        additional_check: Optional callable(chunk_count) -> bool for extra validation
+    """
+    sorted_num_chunks = sorted(num_chunks)
```

Comment:
Does this code handle non-divisible cases ? should we check?

Reply:
Yes, I will add the check `size_before_split % n == 0`